### PR TITLE
Fix another race condition in app-simple

### PR DESCRIPTION
**Pull request by Dretch**
@guibou another concurrency issue :)

`run` failed to finish if an input pipe was still running when the application loop finished. My app had an input that looped forever, emitting an event every 5 seconds, and this prevented the app from exiting.

Now the input pipes are cancelled if they are still going when the application finishes.

### CODE REVIEW

File: gi-gtk-declarative-app-simple/src/GI/Gtk/Declarative/App/Simple.hs [hunk 1]
```diff
@@ -167,24 +168,15 @@ assertRuntimeSupportsBoundThreads = unless rtsSupportsBoundThreads $ do
...
 
-
 publishEvent :: Chan event -> event -> IO ()
 publishEvent mvar = void . writeChan mvar
 
-mergeProducers :: [Producer a IO ()] -> Producer a IO ()
-mergeProducers producers = do
-  (output, input) <- liftIO $ spawn unbounded
-  _               <- liftIO $ Async.mapConcurrently_ (fork output) producers
-  fromInput input
```
Comment (Dretch) [inline gi-gtk-declarative-app-simple/src/GI/Gtk/Declarative/App/Simple.hs:178]: `fromInput` didn't get evaluated until `mapConcurrently_` had returned - so none of the events from the inputs got delivered until all of the events had been produced!

File: gi-gtk-declarative-app-simple/src/GI/Gtk/Declarative/App/Simple.hs [hunk 2]
```diff
@@ -81,24 +83,19 @@ run app = do
...
+  -- If any exception happen in `runLoop`, it will be re-thrown here
   -- and the application will be killed.
-  -- NOTE: the use of Aync.async for `Gtk.main` is *MANDATORY*. The documentation for `
-  -- 'Async.concurrently' says that the first exception 'Async.cancel'
-  -- the second thread only at the condition that it is not a foreign
-  -- call, which is the case for 'Gtk.main'.
-  -- By using a second `Async.async` call, the wrapping thread can be killed.
-  fst <$> Async.concurrently (runLoop app <* Gtk.mainQuit) (Async.async Gtk.main)
+  main <- Async.async Gtk.main
+  runLoop app `finally` (Gtk.mainQuit >> Async.wait main)
```
Comment (Dretch) [inline gi-gtk-declarative-app-simple/src/GI/Gtk/Declarative/App/Simple.hs:89]: This change was necessary to fix a weird intermittent test suite failure. I suspect the issue was either that `Gtk.mainQuit` was not always called (when `runLoop app` threw an exception), or that somehow cancelling `Gtk.main` didn't always work properly.

Comment (guibou) [reply gi-gtk-declarative-app-simple/src/GI/Gtk/Declarative/App/Simple.hs:89]: Damned. Indeed. Sorry about that ;(

Comment (Dretch) [reply gi-gtk-declarative-app-simple/src/GI/Gtk/Declarative/App/Simple.hs:89]: I missed it too :smile:

**Comment by Dretch** [comment]
I have one remaining concurrency concern.

There may be a problem with the threads that run the IO actions from `Transition state action` values that get returned from the `update` function:
- These threads are not cancelled (or awaited) when `run` finishes, unlike everything else.
- Exceptions in these threads are linked to the main `run` thread.
- So asynchronous exceptions could be thrown from the action threads after `run` has finished... which is a bit weird.

I'm not sure if this is really a problem (though I think it is), and I'm not sure what the fix should be. I don't think it should prevent this PR being merged, but I wonder if you have any thoughts @guibou

**Comment by Dretch** [comment]
Github won't allow me to merge this pull request because I created it :smile: 

Only @cblp or @owickstrom have the power to merge it.

However, if they can't do so within a couple of weeks, @guibou perhaps you could recreate the PR yourself -- because then Github will allow me to merge the code ?

**Comment by cblp** [comment]
I'm sorry, I don't have enough expertise to review this.

**Comment by guibou** [comment]
@Dretch I recreated the PR. Feel free to approve it ;)

**Comment by Dretch** [comment]
Thanks @guibou :)

**Pull request merged**